File: lang/java/reef-tests/src/main/java/org/apache/reef/tests/fail/driver/FailDriver.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
package org.apache.reef.tests.fail.driver;

import org.apache.reef.driver.context.*;
import org.apache.reef.driver.evaluator.*;
import org.apache.reef.driver.task.*;
import org.apache.reef.tang.annotations.Name;
import org.apache.reef.tang.annotations.NamedParameter;
import org.apache.reef.tang.annotations.Parameter;
import org.apache.reef.tang.annotations.Unit;
import org.apache.reef.tang.exceptions.BindException;
import org.apache.reef.tests.library.exceptions.DriverSideFailure;
import org.apache.reef.tests.library.exceptions.SimulatedDriverFailure;
import org.apache.reef.wake.EventHandler;
import org.apache.reef.wake.remote.impl.ObjectSerializableCodec;
import org.apache.reef.wake.time.Clock;
import org.apache.reef.wake.time.event.Alarm;
import org.apache.reef.wake.time.event.StartTime;
import org.apache.reef.wake.time.event.StopTime;

import javax.inject.Inject;
import javax.xml.bind.DatatypeConverter;
import java.util.Arrays;
import java.util.logging.Level;
import java.util.logging.Logger;

import static org.apache.reef.tests.fail.driver.FailDriver.ExpectedMessage.RequiredFlag.OPTIONAL;
import static org.apache.reef.tests.fail.driver.FailDriver.ExpectedMessage.RequiredFlag.REQUIRED;

/**
 * Driver which fails on receiving certain message class.
 */
@Unit
public final class FailDriver {

  private static final Logger LOG = Logger.getLogger(FailDriver.class.getName());
  private static final ObjectSerializableCodec<String> CODEC = new ObjectSerializableCodec<>();
  private static final byte[] HELLO_STR = CODEC.encode("MESSAGE::HELLO");
  /**
   * Send message to the Task MSG_DELAY milliseconds after start.
   */
  private static final int MSG_DELAY = 1000;
  private static final ExpectedMessage[] EVENT_SEQUENCE = {
      new ExpectedMessage(FailDriver.class, REQUIRED),
      new ExpectedMessage(StartTime.class, REQUIRED),
      new ExpectedMessage(AllocatedEvaluator.class, REQUIRED),
      new ExpectedMessage(FailedEvaluator.class, OPTIONAL),
      new ExpectedMessage(ActiveContext.class, REQUIRED),
      new ExpectedMessage(ContextMessage.class, OPTIONAL),
      new ExpectedMessage(FailedContext.class, OPTIONAL),
      new ExpectedMessage(RunningTask.class, REQUIRED),
      new ExpectedMessage(Alarm.class, REQUIRED),
      new ExpectedMessage(TaskMessage.class, REQUIRED),
      new ExpectedMessage(Alarm.class, REQUIRED),
      new ExpectedMessage(SuspendedTask.class, REQUIRED),
      new ExpectedMessage(RunningTask.class, REQUIRED),
      new ExpectedMessage(Alarm.class, REQUIRED),
      new ExpectedMessage(FailedTask.class, OPTIONAL),
      new ExpectedMessage(CompletedTask.class, REQUIRED),
      new ExpectedMessage(ClosedContext.class, OPTIONAL),
      new ExpectedMessage(CompletedEvaluator.class, REQUIRED),
      new ExpectedMessage(StopTime.class, REQUIRED)
  };
  private final transient Class<?> failMsgClass;
  private final transient EvaluatorRequestor requestor;
  private final transient Clock clock;
  private transient RunningTask task = null;
  private transient int expectIdx = 0;
  private transient DriverState state = DriverState.INIT;

  @Inject
  public FailDriver(@Parameter(FailMsgClassName.class) final String failMsgClassName,
                    final EvaluatorRequestor requestor, final Clock clock)
      throws ClassNotFoundException {
    this.failMsgClass = ClassLoader.getSystemClassLoader().loadClass(failMsgClassName);
    this.requestor = requestor;
    this.clock = clock;
    this.checkMsgOrder(this);
  }

  /**
   * Check if observer methods are called in the right order
   * and generate an exception at the given point in the message sequence.
   *
   * @param msg a message from one of the observers.
   * @throws SimulatedDriverFailure if failMsgClass matches the message class.
   * @throws DriverSideFailure      if messages are out of order.
   */
  private void checkMsgOrder(final Object msg) throws SimulatedDriverFailure, DriverSideFailure {

    final String msgClassName = msg.getClass().getName();
    LOG.log(Level.FINE, "At {0} {1}:{2}", new Object[]{
        this.state, this.expectIdx, msgClassName});

    if (this.state == DriverState.FAILED) {
      // If already failed, do not do anything
      return;
    }

    // Simulate failure at this step?
    if (this.failMsgClass.isInstance(msg)) {
      this.state = DriverState.FAILED;
    }

    // Make sure events arrive in the right order (specified in EVENT_SEQUENCE):
    boolean notFound = true;
    for (; this.expectIdx < EVENT_SEQUENCE.length; ++this.expectIdx) {
      if (EVENT_SEQUENCE[expectIdx].msgClass.isInstance(msg)) {
        notFound = false;
        break;
      } else if (EVENT_SEQUENCE[expectIdx].requiredFlag == REQUIRED) {
        break;
      }
    }

    if (notFound) {
      LOG.log(Level.SEVERE, "Event out of sequence: {0} {1}:{2}",
          new Object[]{this.state, this.expectIdx, msgClassName});
      throw new DriverSideFailure("Event out of sequence: " + msgClassName);
    }

    LOG.log(Level.INFO, "{0}: send: {1} got: {2}", new Object[]{
        this.state, EVENT_SEQUENCE[this.expectIdx], msgClassName});

    ++this.expectIdx;

    if (this.state == DriverState.FAILED) {
      final SimulatedDriverFailure ex = new SimulatedDriverFailure(
          "Simulated Failure at FailDriver :: " + msgClassName);
      LOG.log(Level.INFO, "Simulated Failure:", ex);
      throw ex;
    }
  }

  private enum DriverState {INIT, SEND_MSG, SUSPEND, RESUME, CLOSE, FAILED}

  /**
   * Name of the message class to specify the failing message handler.
   */
  @NamedParameter(doc = "Full name of the message class to fail on", short_name = "fail")
  public static final class FailMsgClassName implements Name<String> {
  }

  /**
   * Expected message class.
   */
  public static final class ExpectedMessage {

    private final transient Class<?> msgClass;
    private final transient RequiredFlag requiredFlag;
    private final transient String repr;

    public ExpectedMessage(final Class<?> clazz, final RequiredFlag requiredFlag) {
      this.msgClass = clazz;
      this.requiredFlag = requiredFlag;
      this.repr = this.msgClass.getSimpleName() + ":" + this.requiredFlag;
    }

    @Override
    public String toString() {
      return this.repr;
    }

    /**
     * "Required" flag for message class.
     */
    public enum RequiredFlag {OPTIONAL, REQUIRED}
  }

  final class AllocatedEvaluatorHandler implements EventHandler<AllocatedEvaluator> {
    @Override
    public void onNext(final AllocatedEvaluator eval) {
      checkMsgOrder(eval);
      try {
        eval.submitContext(ContextConfiguration.CONF
            .set(ContextConfiguration.IDENTIFIER, "FailContext_" + eval.getId())
            .build());
      } catch (final BindException ex) {
        LOG.log(Level.WARNING, "Context configuration error", ex);
        throw new RuntimeException(ex);
      }
    }
  }

  final class CompletedEvaluatorHandler implements EventHandler<CompletedEvaluator> {
    @Override
    public void onNext(final CompletedEvaluator eval) {
      checkMsgOrder(eval);
      // noop
    }
  }

  final class FailedEvaluatorHandler implements EventHandler<FailedEvaluator> {
    @Override
    public void onNext(final FailedEvaluator eval) {
      LOG.log(Level.WARNING, "Evaluator failed: " + eval.getId(), eval.getEvaluatorException());
      checkMsgOrder(eval);
      throw new RuntimeException(eval.getEvaluatorException());
    }
  }

  final class ActiveContextHandler implements EventHandler<ActiveContext> {
    @Override
    public void onNext(final ActiveContext context) {
      checkMsgOrder(context);
      try {
        context.submitTask(TaskConfiguration.CONF
            .set(TaskConfiguration.IDENTIFIER, "FailTask_" + context.getId())
            .set(TaskConfiguration.TASK, NoopTask.class)
            .set(TaskConfiguration.ON_MESSAGE, NoopTask.DriverMessageHandler.class)
            .set(TaskConfiguration.ON_SUSPEND, NoopTask.TaskSuspendHandler.class)
            .set(TaskConfiguration.ON_CLOSE, NoopTask.TaskCloseHandler.class)
            .set(TaskConfiguration.ON_TASK_STOP, NoopTask.TaskStopHandler.class)
            .set(TaskConfiguration.ON_SEND_MESSAGE, NoopTask.class)
            .build());
      } catch (final BindException ex) {
        LOG.log(Level.WARNING, "Task configuration error", ex);
        throw new RuntimeException(ex);
      }
    }
  }

  final class ContextMessageHandler implements EventHandler<ContextMessage> {
    @Override
    public void onNext(final ContextMessage message) {
      checkMsgOrder(message);
      // noop
    }
  }

  final class ClosedContextHandler implements EventHandler<ClosedContext> {
    @Override
    public void onNext(final ClosedContext context) {
      checkMsgOrder(context);
      // noop
    }
  }

  final class FailedContextHandler implements EventHandler<FailedContext> {
    @Override
    public void onNext(final FailedContext context) {
      LOG.log(Level.WARNING, "Context failed: " + context.getId(), context.getReason().orElse(null));
      checkMsgOrder(context);

      // if (context.getParentContext().isPresent()) {
      //   context.getParentContext().get().close();
      // }
    }
  }

  final class RunningTaskHandler implements EventHandler<RunningTask> {
    @Override
    @SuppressWarnings("checkstyle:hiddenfield")
    public void onNext(final RunningTask task) {
      checkMsgOrder(task);
      FailDriver.this.task = task;
      switch (state) {
      case INIT:
        state = DriverState.SEND_MSG;
        break;
      case RESUME:
        state = DriverState.CLOSE;
        break;
      default:
        LOG.log(Level.WARNING, "Unexpected state at TaskRuntime: {0}", state);
        throw new DriverSideFailure("Unexpected state: " + state);
      }
      // After a delay, send message or suspend the task:
      clock.scheduleAlarm(MSG_DELAY, new AlarmHandler());
    }
  }

  final class SuspendedTaskHandler implements EventHandler<SuspendedTask> {
    @Override
    @SuppressWarnings("checkstyle:hiddenfield")
    public void onNext(final SuspendedTask task) {
      checkMsgOrder(task);
      state = DriverState.RESUME;
      try {
        task.getActiveContext().submitTask(TaskConfiguration.CONF
            .set(TaskConfiguration.IDENTIFIER, task.getId() + "_RESUMED")
            .set(TaskConfiguration.TASK, NoopTask.class)
            .set(TaskConfiguration.ON_MESSAGE, NoopTask.DriverMessageHandler.class)
            .set(TaskConfiguration.ON_SUSPEND, NoopTask.TaskSuspendHandler.class)
            .set(TaskConfiguration.ON_CLOSE, NoopTask.TaskCloseHandler.class)
            .set(TaskConfiguration.ON_TASK_STOP, NoopTask.TaskStopHandler.class)
            .set(TaskConfiguration.ON_SEND_MESSAGE, NoopTask.class)
            .set(TaskConfiguration.MEMENTO, DatatypeConverter.printBase64Binary(HELLO_STR))
            .build());
      } catch (final BindException ex) {
        LOG.log(Level.SEVERE, "Task configuration error", ex);
        throw new DriverSideFailure("Task configuration error", ex);
      }
    }
  }

  final class TaskMessageHandler implements EventHandler<TaskMessage> {
    @Override
    public void onNext(final TaskMessage msg) {
      checkMsgOrder(msg);
      assert Arrays.equals(HELLO_STR, msg.get());
      assert state == DriverState.SEND_MSG;
      state = DriverState.SUSPEND;
      clock.scheduleAlarm(MSG_DELAY, new AlarmHandler());
    }
  }

  final class FailedTaskHandler implements EventHandler<FailedTask> {
    @Override
    @SuppressWarnings("checkstyle:hiddenfield")
    public void onNext(final FailedTask task) {
      LOG.log(Level.WARNING, "Task failed: " + task.getId(), task.getReason().orElse(null));
      checkMsgOrder(task);
      if (task.getActiveContext().isPresent()) {
        task.getActiveContext().get().close();
      }
    }
  }

  final class CompletedTaskHandler implements EventHandler<CompletedTask> {
    @Override
    @SuppressWarnings("checkstyle:hiddenfield")
    public void onNext(final CompletedTask task) {
      checkMsgOrder(task);
      task.getActiveContext().close();
    }
  }

  final class StartHandler implements EventHandler<StartTime> {
    @Override
    public void onNext(final StartTime time) {
      FailDriver.this.checkMsgOrder(time);
      FailDriver.this.requestor.submit(EvaluatorRequest.newBuilder()
          .setNumber(1).setMemory(128).setNumberOfCores(1).build());
    }
  }

  final class AlarmHandler implements EventHandler<Alarm> {
    @Override
    public void onNext(final Alarm time) {
      FailDriver.this.checkMsgOrder(time);
      switch (FailDriver.this.state) {
      case SEND_MSG:
        FailDriver.this.task.send(HELLO_STR);
        break;
      case SUSPEND:
        FailDriver.this.task.suspend();
        break;
      case CLOSE:
        FailDriver.this.task.close();
        break;
      default:
        LOG.log(Level.WARNING, "Unexpected state at AlarmHandler: {0}", FailDriver.this.state);
        throw new DriverSideFailure("Unexpected state: " + FailDriver.this.state);
      }
    }
  }

  final class StopHandler implements EventHandler<StopTime> {
    @Override
    public void onNext(final StopTime time) {
      FailDriver.this.checkMsgOrder(time);
      // noop
    }
  }
}

File: lang/cs/Org.Apache.REEF.Bridge.Core.Tests/Fail/Driver/FailDriver.cs
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Linq;
using Org.Apache.REEF.Bridge.Core.Tests.Fail.ThreadInterruptedException;
using Org.Apache.REEF.Common.Context;
using Org.Apache.REEF.Common.Tasks;
using Org.Apache.REEF.Driver;
using Org.Apache.REEF.Driver.Context;
using Org.Apache.REEF.Driver.Evaluator;
using Org.Apache.REEF.Driver.Task;
using Org.Apache.REEF.Tang.Annotations;
using Org.Apache.REEF.Tang.Exceptions;
using Org.Apache.REEF.Tang.Util;
using Org.Apache.REEF.Utilities;
using Org.Apache.REEF.Utilities.Logging;
using Org.Apache.REEF.Wake.Time;
using Org.Apache.REEF.Wake.Time.Event;

namespace Org.Apache.REEF.Bridge.Core.Tests.Fail.Driver
{
    internal sealed class FailDriver : 
        IObserver<IAllocatedEvaluator>,
        IObserver<ICompletedEvaluator>,
        IObserver<IFailedEvaluator>,
        IObserver<IRunningTask>, 
        IObserver<ICompletedTask>, 
        IObserver<IFailedTask>, 
        IObserver<ITaskMessage>, 
        IObserver<ISuspendedTask>, 
        IObserver<IFailedContext>, 
        IObserver<IClosedContext>, 
        IObserver<IContextMessage>, 
        IObserver<IActiveContext>,
        IObserver<IDriverStarted>,
        IObserver<IDriverStopped>,
        IObserver<Alarm>
    {
        private static readonly Logger Log = Logger.GetLogger(typeof(FailDriver));

        [NamedParameter("fail message type name")]
        internal sealed class FailMsgTypeName : Name<string>
        {
        }

        private enum DriverState
        {
            Init, SendMsg, Suspend, Resume, Close, Failed
        }


        private static readonly byte[] HelloStringByteArray = ByteUtilities.StringToByteArrays("MESSAGE::HELLO");

        private static readonly int MsgDelay = 1000;

        private static readonly ExpectedMessage[] EventSequence = new ExpectedMessage[]
        {
            new ExpectedMessage(typeof(FailDriver), ExpectedMessage.RequiredFlag.Required),
            new ExpectedMessage(typeof(IDriverStarted), ExpectedMessage.RequiredFlag.Required),
            new ExpectedMessage(typeof(IAllocatedEvaluator), ExpectedMessage.RequiredFlag.Required),
            new ExpectedMessage(typeof(IFailedEvaluator), ExpectedMessage.RequiredFlag.Optional),
            new ExpectedMessage(typeof(IActiveContext), ExpectedMessage.RequiredFlag.Required),
            new ExpectedMessage(typeof(IContextMessage), ExpectedMessage.RequiredFlag.Optional),
            new ExpectedMessage(typeof(IFailedContext), ExpectedMessage.RequiredFlag.Optional),
            new ExpectedMessage(typeof(IRunningTask), ExpectedMessage.RequiredFlag.Required),
            new ExpectedMessage(typeof(Alarm), ExpectedMessage.RequiredFlag.Required),
            new ExpectedMessage(typeof(ITaskMessage), ExpectedMessage.RequiredFlag.Required),
            new ExpectedMessage(typeof(Alarm), ExpectedMessage.RequiredFlag.Required),
            new ExpectedMessage(typeof(ISuspendedTask), ExpectedMessage.RequiredFlag.Required),
            new ExpectedMessage(typeof(IRunningTask), ExpectedMessage.RequiredFlag.Required),
            new ExpectedMessage(typeof(Alarm), ExpectedMessage.RequiredFlag.Required),
            new ExpectedMessage(typeof(IFailedTask), ExpectedMessage.RequiredFlag.Optional),
            new ExpectedMessage(typeof(ICompletedTask), ExpectedMessage.RequiredFlag.Required),
            new ExpectedMessage(typeof(ICompletedTask), ExpectedMessage.RequiredFlag.Optional),
            new ExpectedMessage(typeof(ICompletedEvaluator), ExpectedMessage.RequiredFlag.Required),
            new ExpectedMessage(typeof(IDriverStopped), ExpectedMessage.RequiredFlag.Required)
        };

        private static readonly Dictionary<string, Type> TypeHash = new Dictionary<string, Type>()
        {
            {typeof(NoopTask).FullName, typeof(NoopTask)},
            {typeof(FailDriver).FullName, typeof(FailDriver)},
            {typeof(IDriverStarted).FullName, typeof(IDriverStarted)},
            {typeof(IAllocatedEvaluator).FullName, typeof(IAllocatedEvaluator)},
            {typeof(IFailedEvaluator).FullName, typeof(IFailedEvaluator)},
            {typeof(IActiveContext).FullName, typeof(IActiveContext)},
            {typeof(IContextMessage).FullName, typeof(IContextMessage)},
            {typeof(IFailedContext).FullName, typeof(IFailedContext)},
            {typeof(IRunningTask).FullName, typeof(IRunningTask)},
            {typeof(Alarm).FullName, typeof(Alarm)},
            {typeof(ISuspendedTask).FullName, typeof(ISuspendedTask)},
            {typeof(IFailedTask).FullName, typeof(IFailedTask)},
            {typeof(ICompletedTask).FullName, typeof(ICompletedTask)},
            {typeof(ITaskMessage).FullName, typeof(ITaskMessage)},
            {typeof(ICompletedEvaluator).FullName, typeof(ICompletedEvaluator)},
            {typeof(IDriverStopped).FullName, typeof(IDriverStopped)}
        };

        private readonly Type _failMsgClass;
        private readonly IEvaluatorRequestor _requestor;
        private readonly IClock _clock;
        private IRunningTask _task = null;
        private int _expectIdx = 0;
        private DriverState _state = DriverState.Init;

        [Inject]
        public FailDriver([Parameter(Value = typeof(FailMsgTypeName))] string failMsgTypeName,
            IEvaluatorRequestor requestor, IClock clock)
        {
            _clock = clock;
            _requestor = requestor;
            if (TypeHash.ContainsKey(failMsgTypeName))
            {
                _failMsgClass = TypeHash[failMsgTypeName];
            }
            else
            {
                _failMsgClass = typeof(FailDriver).Assembly.GetType(failMsgTypeName, true);
            }
            CheckMsgOrder(this);
        }

        public void OnError(System.Exception error)
        {
            throw new NotImplementedException();
        }

        public void OnCompleted()
        {
            throw new NotImplementedException();
        }

        private void CheckMsgOrder(object obj)
        {
            string msgClassName = _failMsgClass.FullName;
            Log.Log(Level.Info, "Driver state {0} event sequence {1} message obj type {2}", 
                new object[] { _state, EventSequence[_expectIdx].Type, obj.GetType()});

            if (_state == DriverState.Failed)
            {
                // If already failed, do not do anything
                return;
            }

            // Simulate failure at this step?
            if (_failMsgClass.IsInstanceOfType(obj))
            {
                _state = DriverState.Failed;
            }

            // Make sure events arrive in the right order (specified in EVENT_SEQUENCE):
            bool notFound = true;
            for (; _expectIdx < EventSequence.Length; ++_expectIdx)
            {
                if (EventSequence[_expectIdx].Type.IsInstanceOfType(obj))
                {
                    Log.Log(Level.Info, "Object type {0} is instance of expected type {1}", new object[] {obj.GetType(), EventSequence[_expectIdx].Type});
                    notFound = false;
                    break;
                }
                else if (EventSequence[_expectIdx].Flag == ExpectedMessage.RequiredFlag.Required)
                {
                    Log.Log(Level.Info, "Object type {0} is NOT instance of expected type {1}", new object[] {obj.GetType(), EventSequence[_expectIdx].Type});
                    break;
                }
            }

            if (notFound)
            {
                Log.Log(Level.Info, "Event out of sequence: Driver state {0} event sequence {1} message obj type {2}", 
                    new object[] { _state, EventSequence[_expectIdx].Type, obj.GetType()});
                throw new DriverSideFailure("Event out of sequence: " + msgClassName);
            }

            Log.Log(Level.Info, "{0}: send: {1} got: {2}", new object[] {
                _state, EventSequence[_expectIdx], msgClassName});

            ++_expectIdx;

            if (_state == DriverState.Failed)
            {
                SimulatedDriverFailure ex = new SimulatedDriverFailure(
                    "Simulated Failure at FailDriver :: " + msgClassName);
                Log.Log(Level.Info, "Simulated Failure:", ex);
                throw ex;
            }
        }


        private sealed class ExpectedMessage
        {
            public enum RequiredFlag
            {
                Required,
                Optional
            }

            public RequiredFlag Flag { get; }

            public Type Type { get; }

            public ExpectedMessage(Type type, RequiredFlag flag)
            {
                Type = type;
                Flag = flag;
            }
        }

        public void OnNext(IAllocatedEvaluator eval)
        {
            CheckMsgOrder(eval);
            try
            {
                eval.SubmitContext(ContextConfiguration.ConfigurationModule
                    .Set(ContextConfiguration.Identifier, "FailContext_" + eval.Id)
                    .Build());
            }
            catch (BindException ex)
            {
                Log.Log(Level.Warning, "Context configuration error", ex);
                throw new IllegalStateException("context configuration error", ex);
            }
        }

        public void OnNext(ICompletedEvaluator eval)
        {
            CheckMsgOrder(eval);
            // noop
        }

        public void OnNext(IFailedEvaluator eval)
        {
            Log.Log(Level.Warning, "Evaluator failed: " + eval.Id, eval.EvaluatorException);
            CheckMsgOrder(eval);
            throw new IllegalStateException("failed evaluator illegal state", eval.EvaluatorException);
        }

        public void OnNext(IActiveContext context)
        {
            CheckMsgOrder(context);
            try
            {
                context.SubmitTask(TaskConfiguration.ConfigurationModule
                    .Set(TaskConfiguration.Identifier, "FailTask_" + context.Id)
                    .Set(TaskConfiguration.Task, GenericType<NoopTask>.Class)
                    .Set(TaskConfiguration.OnMessage, GenericType<NoopTask>.Class)
                    .Set(TaskConfiguration.OnSuspend, GenericType<NoopTask>.Class)
                    .Set(TaskConfiguration.OnClose, GenericType<NoopTask>.Class)
                    .Set(TaskConfiguration.OnTaskStop, GenericType<NoopTask>.Class)
                    .Set(TaskConfiguration.OnSendMessage, GenericType<NoopTask>.Class)
                    .Build());
            }
            catch (BindException ex)
            {
                Log.Log(Level.Warning, "Task configuration error", ex);
                throw new IllegalStateException("task configuration error", ex);
            }
        }


        public void OnNext(IContextMessage message)
        {
            CheckMsgOrder(message);
            // noop
        }


        public void OnNext(IClosedContext context)
        {
            CheckMsgOrder(context);
            // noop
        }


        public void OnNext(IFailedContext context)
        {
            Log.Log(Level.Warning, "Context failed: " + context.Id);
            CheckMsgOrder(context);
        }


        public void OnNext(IRunningTask runningTask)
        {
            CheckMsgOrder(runningTask);
            _task = runningTask;
            switch (_state)
            {
                case DriverState.Init:
                    Log.Log(Level.Info, "Move to state {0}", DriverState.SendMsg);
                    _state = DriverState.SendMsg;
                    break;
                case DriverState.Resume:
                    Log.Log(Level.Info, "Move to state {0}", DriverState.Close);
                    _state = DriverState.Close;
                    break;
                default:
                    Log.Log(Level.Warning, "Unexpected state at TaskRuntime: {0}", _state);
                    throw new DriverSideFailure("Unexpected state: " + _state);
            }

            // After a delay, send message or suspend the task:
            _clock.ScheduleAlarm(MsgDelay, this);
        }


        public void OnNext(ISuspendedTask suspendedTask)
        {
            CheckMsgOrder(suspendedTask);
            _state = DriverState.Resume;
            try
            {
                suspendedTask.ActiveContext.SubmitTask(TaskConfiguration.ConfigurationModule
                    .Set(TaskConfiguration.Identifier, suspendedTask.Id + "_RESUMED")
                    .Set(TaskConfiguration.Task, GenericType<NoopTask>.Class)
                    .Set(TaskConfiguration.OnMessage, GenericType<NoopTask>.Class)
                    .Set(TaskConfiguration.OnSuspend, GenericType<NoopTask>.Class)
                    .Set(TaskConfiguration.OnClose, GenericType<NoopTask>.Class)
                    .Set(TaskConfiguration.OnTaskStop, GenericType<NoopTask>.Class)
                    .Set(TaskConfiguration.OnSendMessage, GenericType<NoopTask>.Class)
                    .Set(TaskConfiguration.Memento, ByteUtilities.ByteArraysToString(HelloStringByteArray))
                    .Build());
            }
            catch (BindException ex)
            {
                Log.Log(Level.Error, "Task configuration error", ex);
                throw new DriverSideFailure("Task configuration error", ex);
            }
        }

        public void OnNext(ITaskMessage msg)
        {
            CheckMsgOrder(msg);
            Debug.Assert(Enumerable.SequenceEqual(HelloStringByteArray, msg.Message));
            Debug.Assert(_state == DriverState.SendMsg);
            _state = DriverState.Suspend;
            _clock.ScheduleAlarm(MsgDelay, this);
        }

        public void OnNext(IFailedTask failedTask)
        {
            Log.Log(Level.Warning, "Task failed: " + failedTask.Id, failedTask.Reason.OrElse(null));
            CheckMsgOrder(failedTask);
            if (failedTask.GetActiveContext().IsPresent())
            {
                failedTask.GetActiveContext().Value.Dispose();
            }
        }

        public void OnNext(ICompletedTask completedTask)
        {
            CheckMsgOrder(completedTask);
            completedTask.ActiveContext.Dispose();
        }

        public void OnNext(IDriverStarted time)
        {
            CheckMsgOrder(time);
            _requestor.Submit(_requestor.NewBuilder()
                .SetNumber(1).SetMegabytes(128).SetCores(1).Build());
        }

        public void OnNext(Alarm time)
        {
            CheckMsgOrder(time);
            switch (_state)
            {
                case DriverState.SendMsg:
                    Log.Log(Level.Info, "Send message to task {0}", _task.Id);
                    _task.Send(HelloStringByteArray);
                    break;
                case DriverState.Suspend:
                    Log.Log(Level.Info, "Suspend task {0}", _task.Id);
                    _task.Suspend();
                    break;
                case DriverState.Close:
                    Log.Log(Level.Info, "Close task {0}", _task.Id);
                    _task.Dispose();
                    break;
                default:
                    Log.Log(Level.Warning, "Unexpected state at AlarmHandler: {0}", _state);
                    throw new DriverSideFailure("Unexpected state: " + _state);
            }
        }

        public void OnNext(IDriverStopped time)
        {
            CheckMsgOrder(time);
            // noop
        }
    }
}
